### PR TITLE
[State Sync] Add data compression to state sync.

### PR DESCRIPTION
**Pull request by JoshLind**
### Description

This PR adds data compression to state sync (i.e., it utilizes the recently added `aptos-compression` crate: https://github.com/aptos-labs/aptos-core/pull/2232). At a high-level:
1. Storage service clients will specify (in the `StorageServiceRequest`) whether or not the response should be compressed. Compression can be configured via the `AptosDataClientConfig`. The default is true.
2. If the client requests compression, the server will compress the response (see `StorageServiceResponse`).
3. Compression happens at the application layer (instead of at the network layer). There's obviously trade-offs with this approach, but the benefits are: (i) clients can enable/disable compression easily; (ii) the storage service can use cached responses to serve directly to clients, without needing to re-compress data for every request; and (iii) we can selectively compress data, e.g., there's no need to compress all messages, because some are tiny and might actually become larger (like the client requests themselves).

A couple notes:
- Most of this PR is just updating the tests and adding new ones. At the core, we're essentially just making the existing `StorageServiceRequest` and `StorageServiceResponse` messages enums (to support compression).
- This relates to: https://github.com/aptos-labs/aptos-core/issues/541
- We've forked the lz4 and rust-rocksdb crates to workaround temporary issues: https://github.com/rust-rocksdb/rust-rocksdb/issues/666. The forks will be removed once our PRs are accepted upstream 😄 
- (Anecdotal) After running this on a fullnode in devnet, I see a ~4x reduction in network data sizes. State sync throughput was not affected as our bottlenecks are elsewhere in the code. Obviously, mileage in the real world will vary. 🚀 

### Test Plan
New tests have been added, including:
1. Two smoke tests (for validators and fullnodes, testing no compression). All other smoke tests use compression by default.
5. Several unit tests (e.g., to check that the config values work as expected and that client-server interaction is valid).

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2349)

### CODE REVIEW

File: testsuite/smoke-test/src/state_sync_v2.rs
Comment (sitalkedia) [inline testsuite/smoke-test/src/state_sync_v2.rs:258]: Do you want to explicitly enabled compression in `test_full_node_bootstrap_transactions`, so that even if we turn off compression by default, we are still testing both compressed and uncompressed path? Also, possible to reuse the code with above test? :)

**Comment by sitalkedia** [comment]
@JoshLind - Can we summarize the improvements we saw - may be volume of the data as well as the speed of state sync after compression is enabled?

File: state-sync/aptos-data-client/src/aptosnet/tests.rs
Comment (JoshLind) [reply state-sync/aptos-data-client/src/aptosnet/tests.rs:1066]: I actually started doing that but then realized all of these tests need to be cleaned up in very similar ways, so I'd prefer to leave that to a follow up PR (this one is already quite big).

File: testsuite/smoke-test/src/state_sync_v2.rs
Comment (JoshLind) [reply testsuite/smoke-test/src/state_sync_v2.rs:260]: - Cool, I've updated `test_full_node_bootstrap_transactions` and `test_full_node_bootstrap_outputs` to explicitly enable it.
- Regarding re-use, I'm gonna leave that to a follow up exercise :D I think there's a lot more re-use to be had in this file.

File: state-sync/aptos-data-client/src/aptosnet/mod.rs
Comment (JoshLind) [reply state-sync/aptos-data-client/src/aptosnet/mod.rs:507]: Yes, we're currently enabling it for all requests (depending on if it's enabled in the config file). We might want to reconsider this in the future though (e.g., for small response types), but we'll need to see the code in the wild before doing this 🤔

**Comment by JoshLind** [comment]
@sitalkedia:
> Can we summarize the improvements we saw - may be volume of the data as well as the speed of state sync after compression is enabled?

Cool, added a small comment to the PR description, but I didn't want to make it too prominent given that the experimentation is highly anecdotal. I'm much more interested in seeing how this works in the real world 🌏

**Review by sitalkedia — APPROVED** [review]
LGTM

**Comment by sitalkedia** [comment]
@JoshLind - Do you plan to add compression support to consensus as well? I believe consenses messages, particularly proposals can benefit a lot from compression.

**Comment by JoshLind** [comment]
@sitalkedia:
>  Do you plan to add compression support to consensus as well? I believe consenses messages, particularly proposals can benefit a lot from compression.

Yep, already on it and have a working PR 😄 Just need to land this first (need to work with @msmouse to get around a tricky situation with upgrading `rust-rocksdb` and our execution benchmarks).

**Comment by banool** [comment]
@JoshLind, do we have a timeline on when we can cut away from this? In my testing, pulling all the git repos and submodules seems to add (at least locally, though my internet is pretty fast) about 2 minutes to the build.

**Comment by JoshLind** [comment]
> do we have a timeline on when we can cut away from this? In my testing, pulling all the git repos and submodules seems to add (at least locally, though my internet is pretty fast) about 2 minutes to the build.

@banool, do you have more information?
1. If you're asking about compression (generally), I don't see us moving away from this 😄 But, to get this PR to work, we did have to fork 2 crates (one of which is `rust-rocksdb`, which might be the culprit for the additional build time...).
2. We're waiting on outstanding PRs to the upstream crates to land (to kill our forks), but I'm not sure if that'll even help the build time unless we have a breakdown of where the overhead is coming from?

**Pull request merged**
